cmd/skaffold/app/cmd/cmd.go

```diff
@@ -75,26 +75,29 @@ func SetUpLogs(out io.Writer, level string) error {
 	return nil
 }
 
-var defaultConfigForMode = map[bool]*config.SkaffoldConfig{
-	true:  config.DefaultDevSkaffoldConfig,
-	false: config.DefaultRunSkaffoldConfig,
-}
-
 func runSkaffold(out io.Writer, dev bool, filename string) error {
-	buf, err := readConfiguration(filename)
+	buf, err := util.ReadConfiguration(filename)
 	if err != nil {
 		return errors.Wrap(err, "read skaffold config")
 	}
 
-	defaultConfig := defaultConfigForMode[dev]
-	cfg, err := config.Parse(buf, defaultConfig)
+	apiVersion := &config.ApiVersion{}
+	if err := yaml.Unmarshal(buf, apiVersion); err != nil {
+		return errors.Wrap(err, "parsing api version")
+	}
+
+	cfg, err := config.GetConfig(buf, true, dev)
 	if err != nil {
 		return errors.Wrap(err, "parsing skaffold config")
 	}
 
+	if apiVersion.Version != config.LatestVersion {
+		return errors.New("Config version out of date: run `skaffold fix`")
+	}
+
 	opts.Output = out
 	opts.DevMode = dev
-	r, err := runner.NewForConfig(opts, cfg)
+	r, err := runner.NewForConfig(opts, cfg.Config.(*config.SkaffoldConfig))
```

Comment:
The `VersionedConfig` is basically a holder for a `SkaffoldConfig`, which exists so we can easily pair it up with its version without having to explicitly know the type. We can unmarshal any generic config and retrieve its version. Then, if the version is older than `config.LatestVersion`, we know we need to run `skaffold fix`, otherwise we know the config is the "latest" config type, aka `config.SkaffoldConfig` (since we type aliased it there).

Reply:
Since we parse the configs into generic `SkaffoldConfig` objects, we need to type assert this to be the "latest" config before using it. `NewForConfig` uses various fields inside of the config struct (`Build`, `Deploy`, `Artifact`) which might change in the future, so I don't think it really makes sense to expose those via getter methods.

It's safe to type assert here since we already checked the API version to make sure it matches the latest version.